pygmt/base_plotting.py

```diff
@@ -994,3 +994,360 @@ def text(
+                    convention_assigned = True
+                    break
+            if not convention_assigned:
+                raise GMTError(
+                    "Parameters in spec dictionary do not match known " "conventions."
+                )
+
+            # create a dict type pointer for easier to read code
+            if isinstance(spec, dict):
+                dict_type_pointer = list(spec.values())[0]
+            elif isinstance(spec, pd.DataFrame):
+                # use df.values as pointer for DataFrame behavior
+                dict_type_pointer = spec.values
+
+            # assemble the 1D array for the case of floats and ints as values
+            if isinstance(dict_type_pointer, (int, float)):
+                # update pointers
+                set_pointer(data_pointers, spec)
+                # look for optional parameters in the right place
+                lon, lat, depth, plot_lon, plot_lat = update_pointers(data_pointers)
+
+                # Construct the array (order matters)
+                spec = [lon, lat, depth] + [spec[key] for key in foc_params]
+
+                # Add in plotting options, if given, otherwise add 0s as
+                # required by GMT
+                for arg in plot_lon, plot_lat:
+                    if arg is None:
+                        spec.append(0)
+                    else:
+                        if "C" not in kwargs:
+                            kwargs["C"] = True
+                        spec.append(arg)
+
+            # or assemble the 2D array for the case of lists as values
+            elif isinstance(dict_type_pointer, list):
+                # update pointers
+                set_pointer(data_pointers, spec)
+                # look for optional parameters in the right place
+                lon, lat, depth, plot_lon, plot_lat = update_pointers(data_pointers)
+
+                # before constructing the 2D array lets check that each key
+                # of the dict has the same quantity of values to avoid bugs
+                list_length = len(list(spec.values())[0])
+                for value in list(spec.values()):
+                    if len(value) != list_length:
+                        raise GMTError(
+                            "Unequal number of focal mechanism "
+                            "parameters supplied in 'spec'."
+                        )
+                    # lets also check the inputs for lon, lat, and depth if
+                    # it is a list
+                    if (
+                        isinstance(lon, list)
+                        or isinstance(lat, list)
+                        or isinstance(depth, list)
+                    ):
+                        if (len(lon) != len(lat)) or (len(lon) != len(depth)):
+                            raise GMTError(
+                                "Unequal number of focal mechanism "
+                                "locations supplied."
+                            )
```

Comment:
I have a strong feeling that these codes can be greatly simplified into one or two lines by converting everything to numpy array first.

My another concern is that, here you check if lon, lat, and depth are list, but users may also pass 1d numpy arrays to lon, lat, and depth.

Reply:
The purpose of this block is to make sure an array will not be built from lists of unequal lengths. I could do the same thing by making a numpy array, mapping it to the correct size, then searching for missing values but in my opinion that is less readable code. If you have a better idea I'm happy to implement it.

Longitude, latitude, and depth can not be 1d numpy arrays. We specify int float and list types in the docstring.

Reply:
All other PyGMT functions can accept 1d numpy arrays. For example, the *x*, *y* and *sizes* arguments of the `plot` function can be int, float, list or numpy arrays. `meca` should also support it.

Reply:
done